armory/scenarios/scenario.py

```diff
@@ -0,0 +1,411 @@
+            )
+        else:
+            sample_exporter = None
+
+        self.metrics_logger = metrics_logger
+        self.sample_exporter = sample_exporter
+
+    def load(self):
+        self.load_model()
+        if self.use_fit:
+            self.load_train_dataset()
+            self.fit()
+        self.load_attack()
+        self.load_dataset()
+        self.load_metrics()
+        return self
+
+    def evaluate_all(self):
+        logger.info("Running inference on benign and adversarial examples")
+        for _ in tqdm(range(len(self.test_dataset)), desc="Evaluation"):
+            self.next()
+            self.evaluate_current()
+
+    def next(self):
+        x, y = next(self.test_dataset)
+        i = self.i + 1
+        self.i, self.x, self.y = i, x, y
+        self.y_pred, self.y_target, self.x_adv, self.y_pred_adv = None, None, None, None
+
+    def run_benign(self):
+        x, y = self.x, self.y
+        x.flags.writeable = False
+        with metrics.resource_context(name="Inference", **self.profiler_kwargs):
+            y_pred = self.model.predict(x, **self.predict_kwargs)
+        self.metrics_logger.update_task(y, y_pred)
+        self.y_pred = y_pred
+
+    def run_attack(self):
+        x, y, y_pred = self.x, self.y, self.y_pred
+
+        with metrics.resource_context(name="Attack", **self.profiler_kwargs):
+            if self.attack_type == "preloaded":
+                if len(x) == 2:
+                    x, x_adv = x
+                else:
+                    x_adv = x
+                if self.targeted:
+                    y, y_target = y
+                else:
+                    y_target = None
+
+                misclassified = False
+            else:
+                if self.use_label:
+                    y_target = y
+                elif self.targeted:
+                    y_target = self.label_targeter.generate(y)
+                elif self.skip_benign:
+                    y_target = None  # most attacks will call self.model.predict(x)
+                else:
+                    y_target = y_pred
+
+                if self.skip_misclassified:
+                    if self.targeted:
+                        misclassified = all(
+                            metrics.categorical_accuracy(y_target, y_pred)
+                        )
+                    else:
+                        misclassified = not any(metrics.categorical_accuracy(y, y_pred))
+                else:
+                    misclassified = False
+
+                if misclassified:
+                    x_adv = x
+                else:
+                    x_adv = self.attack.generate(
+                        x=x, y=y_target, **self.generate_kwargs
+                    )
+
+        if misclassified:
+            y_pred_adv = y_pred
+        else:
+            # Ensure that input sample isn't overwritten by model
+            x_adv.flags.writeable = False
+            y_pred_adv = self.model.predict(x_adv, **self.predict_kwargs)
```

Comment:
Suggested refactor that I think makes `run_attack()` a bit cleaner:

```suggestion
      with metrics.resource_context(name="Attack", **self.profiler_kwargs):
            if self.attack_type == "preloaded":
                if len(x) == 2:
                    x, x_adv = x
                else:
                    x_adv = x
                if self.targeted:
                    y, y_target = y
                else:
                    y_target = None
                y_pred_adv = y_pred
            else:
                if self.use_label:
                    y_target = y
                elif self.targeted:
                    y_target = self.label_targeter.generate(y)
                elif self.skip_benign:
                    y_target = None  # most attacks will call self.estimator.predict(x)
                else:
                    y_target = y_pred

                if self.skip_misclassified:
                    if self.targeted:
                        misclassified = all(
                            metrics.categorical_accuracy(y_target, y_pred)
                        )
                    else:
                        misclassified = not any(metrics.categorical_accuracy(y, y_pred))
                else:
                    misclassified = None

                if misclassified:
                    x_adv = x
                    y_pred_adv = y_pred
                else:
                    x_adv = self.attack.generate(
                        x=x, y=y_target, **self.generate_kwargs
                    )
                    # Ensure that input sample isn't overwritten by estimator
                    x_adv.flags.writeable = False
                    y_pred_adv = self.estimator.predict(x_adv, **self.predict_kwargs)
```

Reply:
I would like to keep the final `self.estimator.predict` out of the resource_context, so that it's just measuring the cost of generating the perturbed sample.

Generally, though, how do you feel about changing the usage slightly so that it just skips the attack perturbation but doesn't skip the predict step? This would be more realistic in the case of stochastic models.

This would then become:
```
        with metrics.resource_context(name="Attack", **self.profiler_kwargs):            
            if self.attack_type == "preloaded":                                          
                if len(x) == 2:                                                          
                    x, x_adv = x                                                         
                else:                                                                    
                    x_adv = x                                                            
                if self.targeted:                                                        
                    y, y_target = y                                                      
                else:                                                                    
                    y_target = None                                                      
            else:                                                                        
                if self.use_label:                                                       
                    y_target = y                                                         
                elif self.targeted:                                                      
                    y_target = self.label_targeter.generate(y)                           
                elif self.skip_benign:                                                   
                    y_target = None  # most attacks will call self.model.predict(x)      
                else:                                                                    
                    y_target = y_pred                                                    
                                                                                         
                if self.skip_misclassified:                                              
                    if self.targeted:                                                    
                        misclassified = all(                                             
                            metrics.categorical_accuracy(y_target, y_pred)               
                        )                                                                
                    else:                                                                
                        misclassified = not any(metrics.categorical_accuracy(y, y_pred)) 
                else:                                                                    
                    misclassified = False                                                
                                                                                         
                if misclassified:                                                        
                    x_adv = x                                                            
                else:                                                                    
                    x_adv = self.attack.generate(                                        
                        x=x, y=y_target, **self.generate_kwargs                          
                    )                                                                    
                                                                                         
        # Ensure that input sample isn't overwritten by model                            
        x_adv.flags.writeable = False                                                    
        y_pred_adv = self.model.predict(x_adv, **self.predict_kwargs)
```